### PR TITLE
Add star solution in Newtonian gravity

### PR DESCRIPTION
**Pull request by fmahebert**
## Proposed changes

Adds a solution to the Lane-Emden equation, describing a static spherical star in Newtonian Gravity in 3D.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by fmahebert** [comment]
@carmaza Would you be able to review this when you have some time, since you're familiar with the NewtonianEuler code?

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  struct PolytropicConstant {
+    using type = double;
+    static constexpr OptionString help = {
+        "The polytropic constant of the fluid."};
+    static type lower_bound() noexcept { return 0.; }
+  };
+
+  using options = tmpl::list<CentralMassDensity, PolytropicConstant>;
+
+  static constexpr OptionString help = {
```
Comment (carmaza) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp:70]: Can help strings be this long? I believe there are some restrictions, but I can't quite remember where.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+
+def specific_internal_energy(x, t, central_mass_density, polytropic_constant):
+    rho = mass_density(x, t, central_mass_density, polytropic_constant)
+    return polytropic_constant * rho
+
+
+def pressure(x, t, central_mass_density, polytropic_constant):
+    rho = mass_density(x, t, central_mass_density, polytropic_constant)
+    return polytropic_constant * rho**2
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.py:43]: I believe we require at least one empty line at the bottom of py modules, but maybe someone else can confirm.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+SPECTRE_TEST_CASE(
+    "Unit.PointwiseFunctions.AnalyticSolutions.NewtEuler.LaneEmdenStar",
+    "[Unit][PointwiseFunctions]") {
+  pypp::SetupLocalPythonEnvironment local_python_env{
+      "PointwiseFunctions/AnalyticSolutions/NewtonianEuler"};
+
+  test_construction_and_serialization();
+
+  test_solution(std::numeric_limits<double>::signaling_NaN(), 1., 1.);
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp]: If the math is to be independent of the (physically acceptable) values for the central density and the polytropic constant, I'd try less trivial numbers here and in the line below.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  test_solution(DataVector(5), 1., 1.);
+}
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp:109]: For consistency with other initial data, I'd personally add a test catching negative members passed in the input file (with a corresponding assertion for each in the constructor). However, we haven't been super consistent with that, so I'll leave it up to you.

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  if (radius < outer_radius) {
+    const double xi = radius / alpha;
+    return Scalar<double>{{{central_mass_density_ * sin(xi) / xi}}};
+  } else {
+    return Scalar<double>{{{0.0}}};
+  }
+}
+
+template <>
+Scalar<DataVector> LaneEmdenStar::precompute_mass_density(
```
Comment (carmaza) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp]: Possibly useful (i.e. feel free to ignore): if you replace `radius.size()`-> `get_size(radius)` and `get(mass_density)[s]`->`get_element(mass_density, s)` , you can handle both doubles and DataVectors at the same time. These functions are defined in `Utilities/ContainerHelpers.hpp`.

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+ * by solving the Lane-Emden equation \cite Chandrasekhar1939 .
+ * The Lane-Emden equation has closed-form solutions for certain equations of
+ * state; this class implements the solution for a polytropic fluid with
+ * polytropic exponent \f$\Gamma=2\f$ (i.e., with polytropic index \f$n=1\f$).
+ *
+ * The radius and mass of the star are determined by the polytropic constant
+ * \f$\kappa\f$ and central density \f$\rho_c\f$.
+ * The radius is \f$R = \pi \alpha\f$,
+ * and the mass is \f$M = 4 \pi^2 \alpha^3 \rho_c\f$,
+ * where \f$\alpha = \sqrt{\kappa / (2 \pi)}\f$.
```
Comment (carmaza) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp]: with G = 1

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+import numpy as np
+
+
+def mass_density(x, t, central_mass_density, polytropic_constant):
+    alpha = np.sqrt(0.5 * polytropic_constant / np.pi)
+    outer_radius = alpha * np.pi
+    radius = np.sqrt(np.sum(x**2, axis=0)) + 1.e-30 * outer_radius
```
Comment (carmaza) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.py]: From a chat with @osheamonn: you dont need to pass axis as x is a one dimensional array here. Pypp slices tnsr<DataVector> to a tnsr<double> before converting so in the python code, all numpy arrays have the dimensions of the Tensor in the C++ code.

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+namespace NewtonianEuler {
+namespace Sources {
+
+void LaneEmdenGravitationalField::apply(
+    const gsl::not_null<tnsr::I<DataVector, 3>*> source_momentum_density,
+    const Scalar<DataVector>& mass_density,
+    const NewtonianEuler::Solutions::LaneEmdenStar& star,
+    const tnsr::I<DataVector, 3>& x) const noexcept {
+  const auto gravitational_field = star.gravitational_field(x);
+  for (size_t i = 0; i < 3; ++i) {
```
Comment (carmaza) [inline src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:29]: Given a general external acceleration field, the energy equation also requires a source term, which is given by the power density (flow velocity dot field density). For a fluid at rest, though, this source will vanish. Numerically, however, the evolution should _predict_ a vanishing source at later times, as the evolution should predict a vanishing velocity.

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ * The gravitational field \f$g^i\f$ enters the NewtonianEuler system as a
+ * source term for the conserved momentum,
+ *
+ * \f{align*}
+ * \partial_t S^i + \partial_j F^{j}(S^i) = S(S^i) = \rho g^i,
+ * \f}
+ *
+ * where \f$S^i\f$ is the conserved momentum density, \f$F^{j}(S^i)\f$ is the
+ * flux of \f$S^i\f$, \f$\rho\f$ is the fluid mass density.
```
Comment (carmaza) [inline src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp]: It also enters the energy equation (see above).

Note that the energy equation is ignored in the analysis of Lane-Emden stars because the equation is mathematically equivalent to the conservation of entropy density, which renders trivial for isentropic flows such as one described by a polytropic EoS. This is also true for relativistic hydro.

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+ *
+ * \note This source is specialized to the LaneEmdenStar analytic solution in
+ * that it queries a LaneEmdenStar object for the gravitational field. In all
+ * other respects, the source term is generic.
```
Comment (carmaza) [inline src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp]: Regarding this last point, maybe all these sources derived from an external acceleration profile should be somehow handled generically. I have an idea of how to do this but for now I'd suggest to move on with what is done here.

What is important fact to note here, though, is that this source will *not* integrate the density (the one passed as an argument in the apply function) in order to obtain a self-consistent gravitational field (i.e. as if one were solving Euler + Poisson) but rather will return the source generated by a _given_ acceleration field profile (i.e. a function of x), which, in this case, is computed from a Lane-Emden solution. I'd rephrase this note clearly stating all this.

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp [hunk 11]
```diff
@@ -29,6 +29,59 @@ void LaneEmdenStar::pup(PUP::er& p) noexcept {
...
+
+  auto gravitational_field_result = x;
+  for (size_t i = 0; i < 3; ++i) {
+    gravitational_field_result.get(i) *= -enclosed_mass / cube(radius);
+  }
+  return gravitational_field_result;
+}
+
+template <>
+tnsr::I<DataVector, 3> LaneEmdenStar::gravitational_field(
```
Comment (carmaza) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp]: You might also want to use `get_size` and `get_element` (see commit on solution) here if you want to save some lines. Up to you.

File: tests/Unit/Evolution/Systems/NewtonianEuler/Sources/Test_LaneEmdenGravitationalField.cpp [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+            tmpl::list<NewtonianEuler::Tags::MassDensity<DataVector>>{}));
+    this->apply(source_momentum_density, mass_density, star_, x);
+  }
+
+ private:
+  NewtonianEuler::Solutions::LaneEmdenStar star_{};
+};
+
+void test_sources() noexcept {
+  const double central_mass_density = 1.;
```
Comment (carmaza) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Sources/Test_LaneEmdenGravitationalField.cpp]: Try less trivial numbers.

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+#include <cstddef>
+
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp"
+#include "Utilities/Gsl.hpp"
+
+// IWYU pragma: no_include <array>
+
+// IWYU pragma: no_forward_declare Tensor
```
Comment (nilsvu) [inline src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:15]: Have you run IWYU on these files? If no, please remove the pragmas

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+  void pup(PUP::er& /*p*/) noexcept {}  // NOLINT
+
+  using sourced_variables = tmpl::list<Tags::MomentumDensity<DataVector, 3>>;
+
+  using argument_tags = tmpl::list<
+      Tags::MassDensity<DataVector>,
+      ::Tags::AnalyticSolution<NewtonianEuler::Solutions::LaneEmdenStar>,
+      ::Tags::Coordinates<3, Frame::Inertial>>;
+
+  void apply(gsl::not_null<tnsr::I<DataVector, 3>*> source_momentum_density,
```
Comment (nilsvu) [inline src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp]: this function can be static

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp (hunk 11 above)
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp]: I also think that would be good.

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp [hunk 15]
```diff
@@ -0,0 +1,196 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp"
+
+#include <array>
```
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp]: I don't think you use this, but I may have overlooked it

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp (hunk 1 above)
Comment (nilsvu) [reply src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp:70]: They can for option-creatable classes. In this string they can give all the details they want about their options. The help string for individual options is limited in length.

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+  template <typename DataType, typename... Tags>
+  tuples::TaggedTuple<Tags...> variables(const tnsr::I<DataType, 3>& x,
+                                         const double /*t*/,
+                                         tmpl::list<Tags...> /*meta*/) const
+      noexcept {
+    const auto mass_density = precompute_mass_density(x);
+    return {tuples::get<Tags>(variables(tmpl::list<Tags>{}, mass_density))...};
+  }
+
+  /// Compute the gravitational field for the corresponding source term,
```
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.hpp]: Add `\brief`

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.py [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+    if (radius < outer_radius):
+        xi = radius / alpha
+        return central_mass_density * np.sin(xi) / xi
+    else:
+        return 0.
+
+
+def velocity(x, t, central_mass_density, polytropic_constant):
+    return 0. * x
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.py]: np.zeros(3) is a bit better I think

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp [hunk 18]
```diff
@@ -0,0 +1,92 @@
...
+#include "tests/Unit/Pypp/CheckWithRandomValues.hpp"
+#include "tests/Unit/Pypp/SetupLocalPythonEnvironment.hpp"
+#include "tests/Unit/TestCreation.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+namespace {
+
+void test_construction_and_serialization() {
+  const double central_mass_density = 1.;
+  const double polytropic_constant = 1.;
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp]: better use different values to make sure they don't get swapped around

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.py (hunk 2 above)
Comment (fmahebert) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.py:43]: I've browsed through PR #1891, and it looks like the formatting we are about to adopt does _not_ have empty lines at the end of Python files.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp (hunk 4 above)
Comment (fmahebert) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp:109]: This is in tension with the desire to keep the number of tests low (because assertion tests don't combine..) - I am happy to add assertion tests if that is the consensus.

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp (hunk 8 above)
Comment (fmahebert) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:29]: Thanks for catching this 🙂

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp (hunk 10 above)
Comment (fmahebert) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp]: I'll add a note to hopefully make this distinction clearer. I agree it would be nice to bundle the different acceleration-type sources together eventually 🙂

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp (hunk 13 above)
Comment (fmahebert) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:15]: These pragmas are needed

**Comment by fmahebert** [comment]
Thanks for the reviews, @carmaza and @nilsleiffischer! I've posted fixups.

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp [hunk 19]
```diff
@@ -45,19 +45,28 @@ namespace Sources {
...
- * \note This source is specialized to the LaneEmdenStar analytic solution in
- * that it queries a LaneEmdenStar object for the gravitational field. In all
- * other respects, the source term is generic.
+ * \note This source is specialized to the Lane-Emden solution because it
+ * queries a LaneEmdenStar analytic solution for the gravitational field that
+ * generates the fluid acceleration. This source does *not* integrate the fluid
+ * density to compute a self-consistent gravitational field (i.e., as if one
+ * were solving a coupled Euler + Poisson system).
+ *
+ * \note The energy source term is ignored in most analyses of Lane-Emden stars
```
Comment (carmaza) [inline src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp]: I think you can drop this extra note. I only mentioned it to complement what I had mentioned earlier in my review.

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp [hunk 20]
```diff
@@ -19,13 +19,18 @@ namespace Sources {
 
 void LaneEmdenGravitationalField::apply(
     const gsl::not_null<tnsr::I<DataVector, 3>*> source_momentum_density,
-    const Scalar<DataVector>& mass_density,
+    const gsl::not_null<Scalar<DataVector>*> source_energy_density,
+    const Scalar<DataVector>& mass_density_cons,
+    const tnsr::I<DataVector, 3>& momentum_density,
     const NewtonianEuler::Solutions::LaneEmdenStar& star,
-    const tnsr::I<DataVector, 3>& x) const noexcept {
+    const tnsr::I<DataVector, 3>& x) noexcept {
```
Comment (carmaza) [inline src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:26]: Why is this not const? Is it because you changed the function to be static? If I understand correctly, this apply function can be static only because this class has no members, otherwise you'd need it to be non-static.

Source term classes in NewtonianEuler are used in the generic ComputeSources action in the NewtonianEuler directory, which just wraps the call for the particular source's apply member. Note that this generic action _is_ static, as required by the generic evolution code, but the source-specific apply's are not, as the classes are expected to hold members.

**Review by carmaza — COMMENTED** [review]
No extra comments on my side. Feel free to squash.

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp (hunk 19 above)
Comment (fmahebert) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.hpp]: No problem - I'll remove it when I squash

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp (hunk 20 above)
Comment (fmahebert) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:26]: You understand correctly 🙂. I am guessing that @nilsleiffischer suggested to mark this as `static` because then it can be called without instantiating a LaneEmdenGravField object.. but it does reduce consistency with the other NEuler sources. Do you have a preference here?

Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:26]: The way sources are used is by creating an instance in the initial data so, unless we're violating some C++ practice, I'd revert his suggestion in order to keep consistency with the other sources.

Comment (nilsvu) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:26]: You can still call this function the same way as before, on an instance of the class, even if it‘s static. I didn‘t expect that to work either when I first encountered this situation, but it does :) We have been making functions static whenever possible, not sure if there‘s really a performance benefit to it though.

Comment (fmahebert) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:26]: @nilsleiffischer if we're not sure there's a performance benefit, I propose to undo the `static` to maintain consistency with the other sources terms. Ok with you?

Comment (nilsvu) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:26]: I‘m leaning the other way (I think a class shouldn’t have member functions if it has no state; making the function static doesn’t change how you can call it), but it‘s ok for me if you prefer :)

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+    const Scalar<DataType>& mass_density) const noexcept {
+  return make_with_value<tnsr::I<DataType, 3>>(get(mass_density), 0.0);
+}
+
+template <typename DataType>
+tuples::TaggedTuple<Tags::Pressure<DataType>> LaneEmdenStar::variables(
+    tmpl::list<Tags::Pressure<DataType>> /*meta*/,
+    const Scalar<DataType>& mass_density) const noexcept {
+  return Scalar<DataType>{
+      equation_of_state_.pressure_from_density(mass_density)};
```
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/LaneEmdenStar.cpp]: I think the EOS functions return `Scalar`s, so can you remove the `Scalar<DataType>` constructor? (same in function below)

**Review by nilsvu — CHANGES_REQUESTED** [review]
fixups LGTM, please squash (including this change if you make it)

File: src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp (hunk 20 above)
Comment (fmahebert) [reply src/Evolution/Systems/NewtonianEuler/Sources/LaneEmdenGravitationalField.cpp:26]: Ok you've convinced me; I'll keep it static ;)

**Comment by fmahebert** [comment]
Squashed and rebased

**Review by nilsvu — DISMISSED** [review]
@sxs-collaboration/spectre-core-devs This PR is ready for a final look

File: docs/References.bib [hunk 22]
```diff
@@ -158,6 +158,12 @@ @article{Boyle2019kee
       SLACcitation   = "%%CITATION = ARXIV:1904.04831;%%"
 }
 
+@book{Chandrasekhar1939,
+  author = "Chandrasekhar, Subrahmanyan",
+  title  = "An introduction to the study of stellar structure",
+  year   = "1939"
```
Comment (nilsdeppe) [inline docs/References.bib:164]: Another, maybe easier to get a hold of references is Teukolsky and Shapiro

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp (hunk 4 above)
Comment (nilsdeppe) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp:109]: My vote is to not add

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+    return this->variables(x, t, variables_tags<DataType>{});
+  }
+};
+
+template <typename DataType>
+void test_solution(const DataType& used_for_size,
+                   const double central_mass_density,
+                   const double polytropic_constant) noexcept {
+  LaneEmdenStarProxy star(central_mass_density, polytropic_constant);
+  pypp::check_with_random_values<
```
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp:70]: Can you run a serialized and deserialized version through this test too? That'll catch any bugs in serialization and operator==

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+      1, typename LaneEmdenStarProxy::template variables_tags<DataType>>(
+      &LaneEmdenStarProxy::template primitive_variables<DataType>, star,
+      "LaneEmdenStar",
+      {"mass_density", "velocity", "specific_internal_energy", "pressure"},
+      // with polytropic_constant == 2, star has outer radius ~ 1.77
+      {{{-2.0, 2.0}}},
```
Comment (kidder) [inline tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp:76]: don't test with negative quantities that are only physically positive...

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp (hunk 4 above)
Comment (kidder) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp:109]: please test the assertions

File: tests/Unit/Evolution/Systems/NewtonianEuler/Sources/Test_LaneEmdenGravitationalField.cpp [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+  LaneEmdenGravitationalFieldProxy source_proxy(central_mass_density,
+                                                polytropic_constant);
+  pypp::check_with_random_values<3>(
+      &LaneEmdenGravitationalFieldProxy::apply_helper, source_proxy,
+      "Evolution.Systems.NewtonianEuler.Sources.LaneEmdenGravitationalField",
+      {"source_momentum_density", "source_energy_density"},
+      {{{0.0, 3.0},
+        {-1.0, 1.0},
+        // with polytropic_constant == 2, star has outer radius ~ 1.77
+        {-2.0, 2.0}}},
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Sources/Test_LaneEmdenGravitationalField.cpp:58]: positive numbers only for positive quantities

**Review by kidder — CHANGES_REQUESTED** [review]

File: docs/References.bib (hunk 22 above)
Comment (fmahebert) [reply docs/References.bib:164]: But that one does not explicitly give the Gamma=2 (alternatively, n=1) solution that is implemented. I'll use both?

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp (hunk 24 above)
Comment (fmahebert) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp:76]: I don't understand the change request. Here I am trying to sample (x,y,z) coords. Is my `pypp` invocation incorrect?

**Comment by fmahebert** [comment]
Fixups posted, excluding for now the suggestions where (a) I needed further clarification or (b) there is no consensus on path forward.

File: tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp (hunk 24 above)
Comment (kidder) [reply tests/Unit/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/Test_LaneEmdenStar.cpp:76]: never mind, my bad...

File: docs/References.bib (hunk 22 above)
Comment (nilsdeppe) [reply docs/References.bib:164]: Oh, sorry, yes I meant referencing both :)

**Comment by nilsdeppe** [comment]
LGTM (dox run is failing on CI), I'm happy for you to squash once @kidder has taken a look

**Comment by fmahebert** [comment]
Added another fixup to fix the dox error (a silly typo in the new reference)

**Review by kidder — COMMENTED** [review]
squash

**Comment by fmahebert** [comment]
Rebased and squashed

**Review by carmaza — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Review by kidder — APPROVED** [review]
